src/components/pagination/index.tsx

```diff
@@ -0,0 +1,308 @@
+import { useState, createContext, useContext, useMemo } from 'react'
+import {
+	type PropsWithChildren,
+	type Dispatch,
+	type SetStateAction,
+} from 'react'
+import classNames from 'classnames'
+
+import { IconChevronLeft16 } from '@hashicorp/flight-icons/svg-react/chevron-left-16'
+import { IconChevronRight16 } from '@hashicorp/flight-icons/svg-react/chevron-right-16'
+
+import s from './pagination.module.css'
+import { generateTruncatedList } from './helpers'
+
+const PaginationContext = createContext({
+	totalItems: 0,
+	pageSize: 0,
+	setPageSize: (() => void 1) as Dispatch<SetStateAction<number>>,
+	page: 1,
+	setPage: (() => void 1) as Dispatch<SetStateAction<number>>,
+	totalPages: 0,
+	onPageChange: (() => void 1) as (page: number) => void,
+	onPageSizeChange: (() => void 1) as (pagesize: number) => void,
+})
+const usePagination = () => useContext(PaginationContext)
+
+export interface PaginationProps {
+	/** Pass the total number of items to be paginated. If no value is defined an error will be thrown. */
+	totalItems: number
+	/** Pass the maximum number of items to display on each page initially. If no value is defined an error will be thrown. */
+	pageSize: number
+	/**
+	 * Set a custom initial selected page.
+	 *
+	 * Default: `1`
+	 */
+	page?: number
+	/** A callback that fires when the page value changes */
+	onPageChange?: (page: number) => void
+	/** A callback that fires when the page size value changes */
+	onPageSizeChange?: (pagesize: number) => void
+}
+
+const Pagination = ({
+	totalItems,
+	pageSize: _pageSize,
+	page: _page = 1,
+	onPageChange = () => void 1,
+	onPageSizeChange = () => void 1,
+	children,
+}: PropsWithChildren<PaginationProps>) => {
+	if (typeof totalItems !== 'number') {
+		throw new Error(
+			'Pagination: totalItems is required, but was not specified. Please try adding a value such as `103`.'
+		)
+	}
+	if (typeof _pageSize !== 'number') {
+		throw new Error(
+			'Pagination: pageSize is required, but was not specified. Please try adding a value such as `10`.'
+		)
+	}
```

Comment:
Handled in https://github.com/hashicorp/dev-portal/pull/1508/commits/ba1c100e8c70bdea297b1df036d31f2a0780755a

Reply:
💡  / ❓ Would it be possible to return null or `<></>` if the `totalItems` is `0` or _maybe_ `undefined || 0` rather than throwing an error? There are situations when using the pagination on a search page where a search can yield 0 results

@thiskevinwang  @ashleemboyer

Reply:
@pbortnick Thank you and this is a great callout — I'll go and implement this for `totalItems`

Reply:
FWIW I copy pasted in dotcom in the faceted search project and this was the only issue I encountered 🎉 good stuff!

Reply:
@pbortnick: implemented here https://github.com/hashicorp/dev-portal/pull/1508/commits/ca227325be5b3a2fa3b1a2e584b57dfc3eedd8f9

Screenshot
<img width="578" alt="CleanShot 2023-01-11 at 15 50 16@2x" src="https://user-images.githubusercontent.com/26389321/211914972-c6da0b22-71c6-4f79-a46c-a8a2aad553c2.png">

Reply:
That works for me. I actually prefer that to rendering nothing. Thank you 😄 !